cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/ServiceProgramRunner.java

```diff
@@ -124,19 +143,65 @@ public ProgramController run(Program program, ProgramOptions options) {
       // Add a service listener to make sure the plugin instantiator is closed when the http server is finished.
       component.addListener(new ServiceListenerAdapter() {
```

Comment:
Looks like the listener should refactor into a class such that the logic can be shared between `FlowProgramRunner`, `ServiceProgramRunner` and `WorkerProgramRunner`.